File: FinanceTracker.Api/Models/Transaction.cs

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace FinanceTracker.Api.Models
{
    public class Transaction
    {
        public int Id { get; set; }

        [Required, MaxLength(255)]
        public string Description { get; set; } = string.Empty;

        [Required, Column(TypeName = "decimal(10, 2)")]
        public decimal Amount { get; set; }
        public DateTime TransactionDate { get; set; } = DateTime.UtcNow;
        public bool IsExpense { get; set; }

        // Foreign key to Category
        public int CategoryId { get; set; }
        public Category Category { get; set; } = null!;

        // Each transaction belongs to a specific user.
        public int UserId { get; set; }
        public User User { get; set; } = null!;
    }
}

File: finance_tracker_app/lib/models/transaction.dart

class Transaction {
  final int id;
  final String description;
  final double amount;
  final DateTime transactionDate;
  final bool isExpense;
  final int categoryId;
  final Category category;

  Transaction({
    required this.id,
    required this.description,
    required this.amount,
    required this.transactionDate,
    required this.isExpense,
    required this.categoryId,
    required this.category,
  });

  // A factory constructor to create a Transaction from JSON
  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      description: json['description'],
      amount: (json['amount'] as num).toDouble(),
      transactionDate: DateTime.parse(json['transactionDate']),
      isExpense: json['isExpense'],
      categoryId: json['categoryId'],
      category: Category.fromJson(json['category']),
    );
  }
}

class Category {
  final int id;
  final String name;

  Category({required this.id, required this.name});

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      name: json['name'],
    );
  }
}